pkg/operator/ceph/controller/spec.go

```diff
@@ -388,15 +388,19 @@ func AppLabels(appName, namespace string) map[string]string {
 }
 
 // CephDaemonAppLabels returns pod labels common to all Rook-Ceph pods which may be useful for admins.
-// App name is the name of the application: e.g., 'rook-ceph-mon', 'rook-ceph-mgr', etc.
+// App name is the name of the application: e.g., 'rook-ceph-mon', 'rook-ceph-mgr', etc
 // Daemon type is the Ceph daemon type: "mon", "mgr", "osd", "mds", "rgw"
 // Daemon ID is the ID portion of the Ceph daemon name: "a" for "mon.a"; "c" for "mds.c"
-func CephDaemonAppLabels(appName, namespace, daemonType, daemonID string, includeNewLabels bool) map[string]string {
+// BinaryName is the name of the binary running in a container: "ceph-mon", "ceph-mgr", "ganesha.nfsd", etc
+// ResourceName is the resource metadata.name: rook-ceph-mon-a", "rook-ceph-osd-1", etc
+// ResourceKind is the CR type: "CephCluster", "CephFilesystem", etc
+func CephDaemonAppLabels(appName, namespace, daemonType, daemonID, binaryName, resourceName, resourceKind string, includeNewLabels bool) map[string]string {
 	labels := AppLabels(appName, namespace)
 
 	// New labels cannot be applied to match selectors during upgrade
 	if includeNewLabels {
 		labels[daemonTypeLabel] = daemonType
+		k8sutil.AddRecommendedLabels(labels, binaryName, resourceName, resourceKind, daemonType+"-"+daemonID)
```

Comment:
This should be just `daemonID` like I mentioned so that it doesn't risk exceeding the character limit. For CephFilesystem named `my-fs`, this is something like `my-fs-a`, which can be 65 chars if the name of the CephFilesystem is the max 63 chars long.

It will be a little awkward that a mon has `instance: a`, but that is okay for our uses. It is unique in the namespace.

```suggestion
		k8sutil.AddRecommendedLabels(labels, binaryName, resourceName, resourceKind, daemonID)
```

Reply:
Here `Daemon type is the Ceph daemon type: "mon", "mgr", "osd", "mds", "rgw"` is this for file it would be "fs-mirror" which is hardcoded [here](https://github.com/rook/rook/blob/54d532b623610677df66351e18a50ced00b027f8/pkg/operator/ceph/config/config.go#L56).
So I think we can use its a combination.

Reply:
I'm not following. `daemonID` can be longer than 63 characters (which is already an issue, but no one has complained so far), so we should not combine `daemonID` with anything else so we don't make it more likely that we exceed 63 chars.